physical/consul/consul.go

```diff
@@ -240,65 +238,80 @@ func NewConsulBackend(conf map[string]string, logger log.Logger) (physical.Backe
-	default:
-		return nil, err
+func setupTLSConfig(conf map[string]string, tlsConfig *api.TLSConfig, logger log.Logger) (*tls.Config, error) {
+	address, ok := conf["address"]
+	if ok {
+		tlsConfig.Address = address
 	}
 
-	insecureSkipVerify := false
-	if _, ok := conf["tls_skip_verify"]; ok {
-		insecureSkipVerify = true
+	serverName, ok := conf["tls_server_name"]
+	if ok {
+		tlsConfig.Address = serverName
 	}
 
-	tlsMinVersionStr, ok := conf["tls_min_version"]
-	if !ok {
-		// Set the default value
-		tlsMinVersionStr = "tls12"
+	caFile, ok := conf["tls_ca_file"]
+	if ok {
+		tlsConfig.CAFile = caFile
 	}
 
-	tlsMinVersion, ok := tlsutil.TLSLookup[tlsMinVersionStr]
-	if !ok {
-		return nil, fmt.Errorf("invalid 'tls_min_version'")
+	caPath, ok := conf["tls_ca_path"]
+	if ok {
+		tlsConfig.CAPath = caPath
 	}
 
-	tlsClientConfig := &tls.Config{
-		MinVersion:         tlsMinVersion,
-		InsecureSkipVerify: insecureSkipVerify,
-		ServerName:         serverName,
+	certFile, haveCertFile := conf["tls_cert_file"]
+	if haveCertFile {
+		tlsConfig.CertFile = certFile
 	}
 
-	_, okCert := conf["tls_cert_file"]
-	_, okKey := conf["tls_key_file"]
+	keyFile, haveKeyFile := conf["tls_key_file"]
+	if haveKeyFile {
+		tlsConfig.KeyFile = keyFile
+	}
 
-	if okCert && okKey {
-		tlsCert, err := tls.LoadX509KeyPair(conf["tls_cert_file"], conf["tls_key_file"])
-		if err != nil {
-			return nil, fmt.Errorf("client tls setup failed: %v", err)
+	if haveCertFile != haveKeyFile {
+		if haveCertFile {
+			logger.Warn("physical/consul: config tls - missing 'tls_key_file'?")
+		} else {
+			logger.Warn("physical/consul: config tls - missing 'tls_cert_file'?")
 		}
-
-		tlsClientConfig.Certificates = []tls.Certificate{tlsCert}
 	}
 
-	if tlsCaFile, ok := conf["tls_ca_file"]; ok {
-		caPool := x509.NewCertPool()
-
-		data, err := ioutil.ReadFile(tlsCaFile)
+	insecureSkipVerifyStr, ok := conf["tls_skip_verify"]
+	if ok {
+		insecureSkipVerify, err := strconv.ParseBool(insecureSkipVerifyStr)
 		if err != nil {
-			return nil, fmt.Errorf("failed to read CA file: %v", err)
+			return nil, fmt.Errorf("invalid 'tls_skip_verify'")
 		}
+		tlsConfig.InsecureSkipVerify = insecureSkipVerify
+	}
 
-		if !caPool.AppendCertsFromPEM(data) {
-			return nil, fmt.Errorf("failed to parse CA certificate")
-		}
+	logger.Debug("physical/consul: config tls",
+		"Address", tlsConfig.Address,
+		"CAFile", tlsConfig.CAFile,
+		"CAPath", tlsConfig.CAPath,
+		"CertFile", tlsConfig.CertFile,
+		"KeyFile", tlsConfig.KeyFile,
+		"InsecureSkipVerify", tlsConfig.InsecureSkipVerify,
+	)
 
-		tlsClientConfig.RootCAs = caPool
+	tlsClientConfig, err := api.SetupTLSConfig(tlsConfig)
+	if err != nil {
+		return nil, err
 	}
 
+	tlsMinVersionStr, ok := conf["tls_min_version"]
+	if !ok {
+		tlsMinVersionStr = "tls12"
+	}
+
+	tlsMinVersion, ok := tlsutil.TLSLookup[tlsMinVersionStr]
+	if !ok {
+		return nil, fmt.Errorf("invalid 'tls_min_version'")
+	}
+
+	tlsClientConfig.MinVersion = tlsMinVersion
```

Comment:
You did a whole lot of reordering which is making the diff hard to follow. Can you please change the ordering of parsing the various values back to how they were so that it's easier to make sure nothing is being left out?

Reply:
okay, i'll rework the PR to minimize the `diff`